File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        int nv = 0;
        vector<int> id(numCourses);
        vector<vector<int>> adj(numCourses);
        for(int i=0;i<prerequisites.size();i++){
            id[prerequisites[i][0]]++;
            adj[prerequisites[i][1]].push_back(prerequisites[i][0]);
        }
        queue<int>q;
        for(int i=0;i<numCourses;i++){
            if(id[i]==0){
                q.push(i);
            }
        }
        while(!q.empty()){
            int n = q.front();
            q.pop();
            nv++;
            for (auto& neighbor : adj[n]) {
                id[neighbor]--;
                if (id[neighbor] == 0) {
                    q.push(neighbor);
                }
            }
        }
         return nv == numCourses;
    }
};

File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        List<Integer>[] adjacencyList = new ArrayList[numCourses];

        for (int i = 0; i < numCourses; i++) {
            adjacencyList[i] = new ArrayList<>();
        }

        for (int i = 0; i < prerequisites.length; i++) {
            adjacencyList[prerequisites[i][1]].add(prerequisites[i][0]);  
        }

        Set<Integer> visited = new HashSet<>();
        Set<Integer> inStack = new HashSet<>();

        for (int i = 0; i < numCourses; i++) {
            if (!visited.contains(i)) {  
                if (!dfs(i, visited, inStack, adjacencyList)) {  
                    return false;  
                }
            }
        }
        return true;  
    }

    private boolean dfs(int node, Set<Integer> visited, Set<Integer> inStack, List<Integer>[] adjacencyList) {
        visited.add(node);  
        inStack.add(node);  

        List<Integer> neighbors = adjacencyList[node];  

        // Traverse all neighbors
        for (int neighbor : neighbors) {
            if (inStack.contains(neighbor)) {  
                return false;
            }
            if (!visited.contains(neighbor)) {  
                if (!dfs(neighbor, visited, inStack, adjacencyList)) {  
                    return false; 
                }
            }
        }

        inStack.remove(node); 
        return true;  
    }

}